### DISCUSSION TITLE
Connection time too long on Windows

### DISCUSSION THREAD

**Discussion by thep0y** [psycopg/psycopg#190] [Q&A]
```python
import sys
import asyncio
import psycopg

is_windows = sys.platform == "win32"
if is_windows:
    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())


async def test():
    async with await psycopg.AsyncConnection.connect(
        "host=xxxx hostaddr=xxxx user=xxxx password=xxxx dbname=xxxx"
    ) as conn:
        async with conn.cursor() as cur:
            await cur.execute("SELECT id FROM items;")
            row = await cur.fetchone()
            print("id =", row[0])


if __name__ == "__main__":
    import time

    start = time.perf_counter()
    asyncio.run(test())
    print("*" * 60)
    print(f"cost: {time.perf_counter() - start}s")
```

output:
```
id = 1
************************************************************
cost: 5.1114973s
```

The go language driver takes a very short time:
```go
package main

import (
	"database/sql"
	"fmt"
	"time"

	_ "github.com/lib/pq"
)

func main() {
	start := time.Now()

	db, err := sql.Open("postgres", "host=xxxx user=xxxx password=xxxx dbname=xxxx sslmode=disable")
	if err != nil {
		panic(err)
	}

	err = db.Ping()
	if err != nil {
		panic(err)
	}

	row := db.QueryRow("SELECT id FROM items;")
	if row != nil {
		var result uint
		err = row.Scan(&result)
		if err != nil {
			panic(err)
		}
		fmt.Println("id =", result)
	}

	cost := time.Since(start)
	fmt.Println("cost:", cost)
}
```

output:
```
id = 1
cost: 406.0897ms
```

Is there something wrong with my Python code?

**Comment by dvarrazzo** [comment]
Your way of querying the database is wrong: you select an entire table just to use the first result. Psycopg, by default, transfers the entire result to the client, akin to what psycopg does when a server-side cursor is used.

If that's the case, the difference in result is also caused by the query not using any join or sort, so the server starts streaming the result immediately. If there are joins or sort you would be asking the server to do a lot of work just to discard most of it, and using the right query would save a lot of server resources (memory, CPU) that's why the query is "wrong" and not representative of a correct program.

This, of course, ruling out other mistakes, such as connecting to two different databases.

What is the time difference if you add a LIMIT 1 to the query? What if you use a server-side cursor?

For extra curiosity, you can investigate the different behaviour between the two drivers by either logging the server statements or using wireshark.

I also see that you disable SSL in the go driver and not in psycopg. I don't think it makes a huge difference in terms of timing, but if you want to compare the network communication between the two drivers you should do the same (disabling SSL, wireshark can decrypt the Postgres protocol).

**Comment by thep0y** [comment]
> What is the time difference if you add a LIMIT 1 to the query? What if you use a server-side cursor?

I have added "LIMIT 1" to the query, but the time it took was not significantly shorter.
```python
await cur.execute("SELECT id FROM items LIMIT 1;")
```
Windows:
```
id = 30068
************************************************************
cost: 4.663087s
```
The python code is modified based on the example in https://www.psycopg.org/psycopg3/docs/advanced/async.html, so I didn’t realize that the way of querying was wrong.

> I also see that you disable SSL in the go driver and not in psycopg. I don't think it makes a huge difference in terms of timing, but if you want to compare the network communication between the two drivers you should do the same (disabling SSL, wireshark can decrypt the Postgres protocol).

If SSL is not disabled, `github.com/lib/pq` will panic, so SSL must be disabled. I also think that disabling SSL has little effect on time consumption, because the same python code takes a very short time on the Linux platform.
Linux:
```
id = 30068
************************************************************
cost: 0.09465709999949468s
```

**Comment by dvarrazzo** [comment]
> cost: 4.663087s

4 seconds is still a ridiculously long time. Where is the time spent? In the connection or in the query? I think you have some networking problem there.

> The python code is modified based on the example in https://www.psycopg.org/psycopg3/docs/advanced/async.html, so I didn’t realize that the way of querying was wrong.

The query is just an example. If your test table has one record it's fine. If the query has 1M record it's stupid. If the query is in a real program it's stupid. The psycopg documentation is not the right place to learn sql or postgres.

> If SSL is not disabled, github.com/lib/pq will panic, so SSL must be disabled.

lol, sounds safe.

But yeah, makes me think that the driver has its own protocol parser. It receives the first record and cuts the communication with the server. The result is the connection is probably unusable from there after. You would probably find an error of unclean disconnection on the server.

Closing the ticket because that's an issue with your window network, not with psycopg.

**Comment by dvarrazzo** [comment]
It might even well be that the 4s you see on windows are taken by the ssl library to establish the connection. Make a test with `sslmode=disable` with psycopg too.

**Comment by thep0y** [comment]
Change the code to sync, and disable SSL:
```python
def main():
    point1 = time.perf_counter()
    conn: Connection = psycopg.connect(
        "hostaddr=xxxx user=xxxx password=xxxx dbname=xxxx sslmode=disable"
    )
    point2 = time.perf_counter()
    print("connection time:", point2 - point1)
    print("*" * 60)
    cur = conn.cursor()
    cur.execute("SELECT id FROM items LIMIT 1;")
    row = cur.fetchone()
    print(f"result: id={row[0]}")
    point3 = time.perf_counter()
    print("*" * 60)
    print("query time:", point3 - point2)
```
---
output on Windows:
```
connection time: 4.6051261
************************************************************
result: id=30068
************************************************************
query time: 0.024725700000000295
```

output on WSL2 Linux:
```
connection time: 0.0460120999996434
************************************************************
result: id=30068
************************************************************
query time: 0.01619459999983519
```
---
In the Windows platform, the connection takes too long, even if the `hostaddr` is used directly and SSL is disabled.
But in the WSL 2 Linux(Ubuntu), the time it takes to connect will be very short.

Windows and WSL2 use the same network, but the time it takes to connect is different.

---

Connect to `MySQL` on the same server:
```python
def main():
    point1 = time.perf_counter()
    conn = pymysql.connect(
        host="xxxx", user="xxxx", passwd="xxxx", database="xxxx"
    )
    point2 = time.perf_counter()
    print("connection time:", point2 - point1)
    print("*" * 60)
    cur = conn.cursor()
    cur.execute("SELECT id FROM items LIMIT 1;")
    row = cur.fetchone()
    print(f"result: id={row[0]}")
    point3 = time.perf_counter()
    print("*" * 60)
    print("query time:", point3 - point2)
```
output on Windows:
```
connection time: 0.043309099999999996
************************************************************
result: id=1
************************************************************
query time: 0.006753800000000004
```
The time it takes to connect to MySQL is normal, so it doesn't look like a network issue in Windows.

**Comment by dvarrazzo** [comment]
You can get a better insight about what happens during the connection by logging the status during connection attempt: here, for example:

https://github.com/psycopg/psycopg/blob/e2458b9f59c8e8741684ce14979f304c206b5f90/psycopg/psycopg/generators.py#L39

but make sure that you use the Python implementation of psycopg (e.g. by setting `set PSYCOPG_IMPL=python`) otherwise that function won't be called. For instance adding this log line:

```diff
--- a/psycopg/psycopg/generators.py
+++ b/psycopg/psycopg/generators.py
@@ -36,6 +36,7 @@ def connect(conninfo: str) -> PQGenConn[PGconn]:
     """
     conn = pq.PGconn.connect_start(conninfo.encode())
     while 1:
+        logger.info("status: %s", ConnStatus(conn.status).name)
         if conn.status == ConnStatus.BAD:
             raise e.OperationalError(
                 f"connection is bad: {pq.error_message(conn)}"
```

You can get a logging trace similar to the following, Make sure to use the right Windows way to set the env var - this example is on Linux:

```
$ PSYCOPG_IMPL=python ipython

In [1]: import logging
In [2]: logging.basicConfig(format="%(asctime)s %(message)s")
In [3]: logging.getLogger("psycopg.generators").setLevel(logging.INFO)
In [4]: import psycopg
In [5]: psycopg.connect("host=localhost")

2021-12-28 02:58:13,463 status: STARTED
2021-12-28 02:58:13,463 status: MADE
2021-12-28 02:58:13,464 status: SSL_STARTUP
2021-12-28 02:58:13,468 status: SSL_STARTUP
2021-12-28 02:58:13,471 status: SSL_STARTUP
2021-12-28 02:58:13,472 status: MADE
2021-12-28 02:58:13,472 status: AWAITING_RESPONSE
2021-12-28 02:58:13,474 status: AWAITING_RESPONSE
Out[5]: <psycopg.Connection [IDLE] (host=localhost database=piro) at 0x7f9e4c8890d0>
```

You can try and spot the difference between connecting on Windows, Linux, and WLS.

**Comment by dvarrazzo** [comment]
Converted to discussion, because it's not a psycopg bug but it might highlight some interesting result on windows.

**Reply by thep0y** [reply]
```
2021-12-28 11:07:13,826 status: STARTED
2021-12-28 11:07:13,837 status: MADE
2021-12-28 11:07:13,854 status: GSS_STARTUP
2021-12-28 11:07:18,400 status: STARTED
2021-12-28 11:07:18,400 status: MADE
2021-12-28 11:07:18,409 status: AWAITING_RESPONSE
2021-12-28 11:07:18,409 status: AWAITING_RESPONSE
2021-12-28 11:07:18,424 status: AWAITING_RESPONSE
```

The issue comes from `gssencmode`, it is normal when I add `gssencmode=disable` to connect

**Reply by dvarrazzo** [reply]
Good finding. It's not the first time I have seen people having to disable GSS on Windows. You should hear from the PostgreSQL people why that's the case. You can try and write the pgsql-general mailing list for instance.
